<a href="https://colab.research.google.com/github/XimenaBG/XimenaBG.github.io/blob/master/Guided_Project_Exploring_Hacker_News_Posts.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploring Hacker News Posts** 

*The aim of this project is to explore the Ask HN and Show HN posts to determine which one receive more comments on average and if posts created at a certain time receive more comments on average.*


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
from csv import reader

opened_file = open('/content/drive/MyDrive/Colab Notebooks/hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
print(hn[:5])
headers = hn[0]
hn = hn[1:]
print('\n')
print(headers)
print('\n')
print(hn[:5])
print('\n')
print(len(hn))

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0ph

In [ ]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
  title = row[1]
  title = title.lower()
  if title.startswith('ask hn'):
    ask_posts.append(row)
  elif title.startswith('show hn'):
    show_posts.append(row)
  else:
    other_posts.append(row)

print('Number of Ask Posts:', len(ask_posts))
print('Number of Show Posts:', len(show_posts))
print('Number of Other Posts:', len(other_posts))


Number of Ask Posts: 1744
Number of Show Posts: 1162
Number of Other Posts: 17194


In [ ]:
total_ask_comments = 0
for row in ask_posts:
  num_comments = row[4]
  total_ask_comments += int(num_comments)
  avg_ask_comments = total_ask_comments/len(ask_posts)
print('Average ask comments:', avg_ask_comments)

total_show_comments = 0
for row in show_posts:
  num_comments_s = row[4]
  total_show_comments += int(num_comments_s)
  avg_show_comments = total_show_comments/len(show_posts)
print('Average show comments:',avg_show_comments)

Average ask comments: 14.038417431192661
Average show comments: 10.31669535283993


*As we can see in the results the average ask comments is greater than the average show comments. 
This behaviour can be explained by the fact that people likes to be useful and helpful and when somebody asks a question and you know the response you want to offer your help. On the other hand the show posts where you can share something new or innovative or simply relevant to the community maybe do not receive the same ammount of interactions because people do not feel the same engagement like when they are bein helpful.* 

In [ ]:
import datetime as dt

result_list =[]
for row in ask_posts:
  created_at = row[6]
  n_comments = row[4]
  n_comments = int(n_comments)
  result_list.append([created_at, n_comments])

print('Number of Ask Posts:', len(result_list))
print('Number of comments created at:', result_list[:5])
print('\n')

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
  created = row[0]
  # in the following line first of all i create the datetime object (strptime) and after that i just select the hour from the object (strftime)
  hour = dt.datetime.strptime(created, date_format).strftime("%H") 
  if hour not in counts_by_hour:
    counts_by_hour[hour] = 1
    comments_by_hour[hour] = row[1]
  else:
    counts_by_hour[hour] += 1
    comments_by_hour[hour] += row[1]  

print('Ask posts created in each hour:', counts_by_hour)
print('\n')
print('Comments created in each hour:', comments_by_hour)


Number of Ask Posts: 1744
Number of comments created at: [['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


Ask posts created in each hour: {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


Comments created in each hour: {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [ ]:
comments = []
for hour in comments_by_hour:
  avg_by_hour = comments_by_hour[hour]/counts_by_hour[hour]
  comments.append([hour, avg_by_hour])

print(comments)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [ ]:
swap_avg_by_hour = []
for row in comments:
  swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)
print('\n')

sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap)
print('\n')

print("Top 5 hours for ask post comments:")

hour_format = "%H"
for row in sorted_swap[:5]:
  created = row[1]
  hour = dt.datetime.strptime(created, hour_format).strftime("%H:%M") 
  value = row[0]
  string = "{hour_f} : {value_f:.2f} average comments per post"
  output = string.format(hour_f = hour, value_f = value)
  print(output)



[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'],

*As we can see the best hour to create an ask post is at 15:00, the average comments per post at this time is 38.59.
The date and time of the posts in this dataset was made in the time zone  Eastern Time in the US, and now I am livinging in Brazil specifically in the southeast for that reason we could write at 17:00 to take advantage of the engagement of the users at this time.* 

Next steps:
1. Determine if show or ask posts receive more points on average.
2. Determine if posts created at a certain time are more likely to receive more points.
3. Compare your results to the average number of comments and points other posts receive.
4. Use Dataquest's data science project style guide to format your project.